Add Support for building and booting multicore images on LPC54114

galak opened the pull request:
This PR is based on the work done in #7161.  It reworks how we include an image and load it for the second core.

stanislav-poboril committed:
mcux: Add MCUX IPM driver for lpc and kinetis socs

Add driver for MCUX mailbox which can be used for lpcxpresso54114
and other lpc and kinetis socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
sample: Add MCUX IPM sample application

Add MCUX IPM sample application. It can be run on lpcxpresso54114
board at the moment.

We first build the slave core image out of the remote/ dir than the
primary core image is build which includes the slave core image.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

stanislav-poboril committed:
arm: lpcxpresso54114_m0: Add board support for slave core

Add a board port to allow creation of "firmware" code that will run on
the slave core (Cortex-M0+).

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
arch: arm: lpc: Added support for Cortex-M0+ on lpc54114 soc

Add soc configuration support and dts files for nxp_lpc54xxx_m0.

Adjusted nxp_lpc54xxx soc, configuration and dts files for the
presence of slave core.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
arch: arm: nxp_lpc: Added support for init of slave core

Added ability to load and bootup the slave (Cortex-M0+) core on the
LPC54xxx SoC.  Currently we expect a binary image to be specified via
Kconfig for the code that will run on the slave core.  We also specify
via Kconfig the boot/load address for the slave core.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@nashif can you look at the latest commit that adds a sample.  Not sure what the latest is with what we need / want in sample.yaml.

stanislav-poboril commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.soc:
```diff
@@ -35,4 +35,23 @@ config SOC_PART_NUMBER_LPC54XXX
+	depends on SLAVE_CORE_MCUX
+	string "Binary image of slave core's code"
+
+config SLAVE_BOOT_ADDRESS
```
Should SLAVE_BOOT_ADDRESS depend on SLAVE_CORE_MCUX? Or is the intention to be able to use this on slave core too? (currently not used by slave core)

galak replied:
It should depend on SLAVE_CORE_MCUX

stanislav-poboril reviewed (approved):
Thanks.

galak committed:
cmake: Add new generate_inc_file_for_gen_target function

Add a variant on generate_inc_file_for_target in which the user can
specify the target that the generated file will depend on.  This allows
exposing that target to other possible places where a dependency on the
generated file existing might be needed.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

dbkinder commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,61 @@
+.. _lpcxpresso54114_m0:
+
+NXP LPCXPRESSO54114
```
Also noticed PR #7560 is also updating this doc

dbkinder replied:
sorry, meant PR #7416 (but that looks like a rogue PR).

nashif commented on samples/subsys/ipc/ipm_mcux/master/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+sample:
+    description: TBD
+    name: TBD
```
replace TBD with something useful :)

nashif commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.soc:
```diff
@@ -30,4 +35,24 @@ config SOC_PART_NUMBER_LPC54XXX
+config SLAVE_IMAGE
+	depends on SLAVE_CORE_MCUX
+	string "Binary image of slave core's code"
+
```
add help section.

nashif commented on samples/subsys/ipc/ipm_mcux/master/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+set(BOARD lpcxpresso54114_m4)
+include(ExternalProject)
+
+ExternalProject_Add(
```
interesting, is this going to be the way we build multi-core images? Can we have an RFC and eventually document this somewhere and streamline it a bit?

nashif replied:
How about having 'remote' project be part of the main  project, so you end up with 

```
ipm_mcux/
     CMakeLists.txt
     prj.conf
      master/*.c
      remote/
          prj.conf
          CMakeLists.txt
```

galak replied:
> interesting, is this going to be the way we build multi-core images? Can we have an RFC and eventually document this somewhere and streamline it a bit?

Do we have an issue related to multi-core images, I think this is the start of how we would want to do certain aspects of it.

galak replied:
> How about having 'remote' project be part of the main project, so you end up with

I've made this change, makes sense to me.

nashif replied:
> Do we have an issue related to multi-core images, I think this is the start of how we would want to do certain aspects of it.

we have this https://github.com/zephyrproject-rtos/zephyr/issues/3066, but we need to expand the problem description.

nashif commented on samples/subsys/ipc/ipm_mcux/README.rst:
```diff
@@ -0,0 +1,43 @@
+Building and Running
+********************
+
+.. zephyr-app-commands::
```
we have no CMakeLists.txt file here, so this will not work, will it?

galak replied:
With re-org of the dir, this should be ok now.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.soc:
```diff
@@ -30,4 +35,24 @@ config SOC_PART_NUMBER_LPC54XXX
+	help
+	  Driver for slave core startup
+
+config SLAVE_IMAGE
```
Hard to say, will add _MCUX to the symbol for now.  We need a few other multicore devices and re-look at how we deal with multiimage creation.  But will leave that for the future.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.soc:
```diff
@@ -30,4 +35,24 @@ config SOC_PART_NUMBER_LPC54XXX
+	depends on SLAVE_CORE_MCUX
+	string "Binary image of slave core's code"
+
+config SLAVE_BOOT_ADDRESS
```
Added MCUX.

galak commented on samples/subsys/ipc/ipm_mcux/remote/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+    description: TBD
```
done

MaureenHelm commented on drivers/ipm/ipm_mcux.c:
```diff
@@ -0,0 +1,176 @@
+
+	cpu_id = MAILBOX_ID_THIS_CPU;
+
+	volatile u32_t value = MAILBOX_GetValue(MAILBOX, cpu_id);
```
Get the base address from the device config. There are a few other places in the driver that need the same change

galak replied:
Done, fixed to use base from config struct

MaureenHelm commented on drivers/ipm/ipm_mcux.c:
```diff
@@ -0,0 +1,176 @@
+static void mcux_mailbox_config_func_0(struct device *dev);
+
+static const struct mcux_mailbox_config mcux_mailbox_0_config = {
+	.base = MAILBOX,
```
Use the base address from dts

galak replied:
Done

MaureenHelm commented on samples/subsys/ipc/ipm_mcux/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+set(BOARD lpcxpresso54114_m4)
```
Add a license header and copyright. Do this for all the new files

galak replied:
Fixed

nashif commented on samples/subsys/ipc/ipm_mcux/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+ExternalProject_Add(
+  ipm_mcux_remote
+  SOURCE_DIR $ENV{ZEPHYR_BASE}/samples/subsys/ipc/ipm_mcux/remote
```
```
-  SOURCE_DIR $ENV{ZEPHYR_BASE}/samples/subsys/ipc/ipm_mcux/remote
+  SOURCE_DIR ${APPLICATION_SOURCE_DIR}/remote
```

galak replied:
Figured it out.

MaureenHelm commented on samples/subsys/ipc/ipm_mcux/remote/src/main_remote.c:
```diff
@@ -0,0 +1,32 @@
+
+void main(void)
+{
+	ipm = device_get_binding("MAILBOX_0");
```
Use MAILBOX_0_LABEL instead of hardcoding the string

galak replied:
fixed

MaureenHelm commented on samples/subsys/ipc/ipm_mcux/src/main_master.c:
```diff
@@ -0,0 +1,52 @@
+	printk("Hello World from MASTER! %s\n", CONFIG_ARCH);
+
+	/* Get IPM device handle */
+	ipm = device_get_binding("MAILBOX_0");
```
Use MAILBOX_0_LABEL instead of hardcoding the string

galak replied:
fixed

galak commented on ext/hal/nxp/mcux/CMakeLists.txt:
```diff
@@ -5,7 +5,12 @@ string(TOUPPER ${CONFIG_SOC} MCUX_DEVICE)
   set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER}_cm4)
 else()
-  set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER})
+  if("${MCUX_DEVICE}" STREQUAL "LPC54114_M0")
```
DOne

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.defconfig.lpc54114_m0:
```diff
@@ -0,0 +1,42 @@
+	string
+	default lpc54114_m0
+
+config CPU_CORTEX_M_HAS_VTOR
```
Fixed

MaureenHelm commented on ext/hal/nxp/mcux/CMakeLists.txt:
```diff
@@ -4,8 +4,11 @@ string(TOUPPER ${CONFIG_SOC} MCUX_DEVICE)
+  set(MCUX_DEVICE LPC54114)
 else()
-  set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER})
+    set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER})
```
nit: this line shouldn't change

galak replied:
fixed

MaureenHelm commented on samples/subsys/ipc/ipm_mcux/README.rst:
```diff
@@ -0,0 +1,43 @@
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/ipc/ipm_mcux
+   :board: lpcxpresso54114
```
Need to append `_M4` since the board was renamed

galak replied:
fixed

MaureenHelm reviewed (commented):
Just a couple of nits
